src/routers/alpha-router/alpha-router.ts

```diff
@@ -1471,7 +1471,60 @@ export class AlphaRouter
     const V3poolsRaw = V3poolAccessor.getAllPools();
     const V2poolsRaw = V2poolAccessor.getAllPools();
 
-    const poolsRaw = [...V3poolsRaw, ...V2poolsRaw];
+    /// not just [Protocol.MIXED]
+    let chosenV2Pools: Pair[] = [];
+    if (
+      !routingConfig.forceMixedRoutes &&
+      routingConfig.protocols !== undefined &&
+      routingConfig.protocols.length > 1
+    ) {
+      /// Only consider V2 pools if amounts[100] (full amountIn) is of significant size compared to pool TVL
+      /// start out very conservative, let's do like 0.5%
+      const amountIn = amounts[amounts.length - 1]!;
+      /// TODO: direct swap pool is usually 10_000 (does not exist) so we could remove that
+      V2poolsRaw.forEach((pool) => {
+        if (
+          pool.involvesToken(amountIn.currency.wrapped) &&
+          amountIn.greaterThan(
+            pool.reserveOf(amountIn.currency.wrapped).divide(2000) // 0.005%
+          )
+        ) {
+          log.info(
+            {
+              pool: pool.toString(),
+              amountIn: amountIn.toString(),
+              reserveAmountIn: pool
+                .reserveOf(amountIn.currency.wrapped)
+                .toString(),
+            },
+            'AmountHeuristic[AmountIn]: Considering V2 pool for mixed routes'
+          );
+          chosenV2Pools.push(pool);
+        }
+        if (
+          pool.involvesToken(quoteToken) &&
+          amountIn.greaterThan(
+            pool.reserveOf(quoteToken).divide(2000) // 0.005%
+          )
+        ) {
+          log.info(
+            {
+              pool: pool.toString(),
+              amountIn: amountIn.toString(),
+              reserveQuoteToken: pool.reserveOf(quoteToken).toString(),
+            },
+            'AmountHeuristic[AmountIn]: Considering V2 pool for mixed routes'
+          );
+          chosenV2Pools.push(pool);
+        }
+      });
+    }
+
+    log.info(
+      `Chose ${chosenV2Pools.length} out of ${V2poolsRaw.length} V2 pools for mixed routes`
+    );
+
+    const poolsRaw = [...V3poolsRaw, ...chosenV2Pools];
```

Comment:
`chosenV2Pools` is empty if the condition above is not met. what if `routingConfig.protocols.length = 1` and it's just mixed routes. We still want to be looking at v2 pools.

Reply:
If we're only querying for `[Protocols.MIXED]` we would always want to return a mixed route if a route does exist. Using this heuristic, we only return mixed routes that we think are good. 

I guess there should be another conditional case where we use `...V2poolsRaw` instead of `...chosenV2Pools`